Add support for passing custom data to context through `App.Process()`

singhk97 opened the pull request:
https://github.com/microsoft/teams.net/issues/81

heyitsaamir reviewed (dismissed):
Tests?

aacebo commented on Libraries/Microsoft.Teams.Apps/AppBuilder.cs:
```diff
@@ -95,6 +95,12 @@ public AppBuilder AddOAuth(string defaultConnectionName)
         return this;
     }
 
+    public AppBuilder AddContextExtra(IDictionary<string, object> extra)
```
this is nice for global scenario, but I think they also want to be able to pass in custom data on each request, meaning we would need to also support this on a per request basis in the aspnetcore plugin and the `App.Process` method

singhk97 replied:
good catch - pushed commit

lilyydu commented on Libraries/Microsoft.Teams.Apps/App.cs:
```diff
@@ -250,6 +252,19 @@ public async Task<Response> Process(ISenderPlugin sender, IToken token, IActivit
+            {
+                foreach (var kvp in plugin.ContextExtra)
+                {
+                    mergedExtra[kvp.Key] = kvp.Value;
```
if there was already an entry, this would override right? should we add that as a note somewhere?

singhk97 replied:
once this PR gets merged i'll write some docs for this feature (possibly in [this](https://microsoft.github.io/teams-ai/typescript/essentials/app-basics#plugins) section) - as for this i'll add a comment

lilyydu reviewed (commented):
should we also add docs for this?

heyitsaamir commented on Libraries/Microsoft.Teams.Apps/Plugins/Plugin.cs:
```diff
@@ -8,6 +8,8 @@ namespace Microsoft.Teams.Apps.Plugins;
 /// </summary>
 public interface IPlugin
 {
+    public IDictionary<string, object>? ContextExtra { get; set; }
```
should this be a `buildContextExtra`? since it might be different for each execution?

lilyydu replied:
hmm if its per request, maybe we can somehow denote that?

for the `extra` part`, it might be a little trippy where extra denotes "additional" but it can also override old entries

aacebo replied:
left a comment but this should be on a request basis and plugins are initialized on startup, we don't want things to be global we want to add a way to set this extra key/value data on a per-request basis via the existing `Do` sender method https://github.com/microsoft/teams.net/blob/main/Libraries/Microsoft.Teams.Apps/Plugins/SenderPlugin.cs#L45 (which should be an overload since some may not want to set that extra data)

aacebo reviewed (changes requested):
a few comments but good direction

rajan-chari commented:
If there are multiple instances of an App w/ round-robin load balancing, what's the expectation from a dev perspective about lifetime of Context-Extra?  For example, if auth happens in the middle of servicing an activity?

singhk97 commented:
> If there are multiple instances of an App w/ round-robin load balancing, what's the expectation from a dev perspective about lifetime of Context-Extra? For example, if auth happens in the middle of servicing an activity?

The ContextExtra is a per-request level object. That means that it should be created on the fly when the app recieves an incomming request. And be disposed once the app has handled the request. This includes if auth happens in the middle of servicing an activity.

singhk97 commented:
Latest update based on @aacebo's feedback:

* Removed the `IPlugin.ContextExtra` property since it introduces state where it's not needed. The context extra is meant to be a per-request level object - and adding it as a property can cause uninteded side effects such as the object perisisting beyond the current turn.
* Modified the `ISenderPlugin.Do` to take in context extra as a paramter since that's the entry-point to the app's activity processing method.

aacebo reviewed (commented):
can you also add a test case for setting a new key/value from a handler and asserting that key/value being present in the next chained handler?

singhk97 commented:
For future reference:

@aacebo:
> plugins were designed to be out of sight out of mind, and I specifically made the decision to put context creation on the app not the plugins for several reasons, one of which being that even tho one plugin may receive the activity, any plugin can handle that activity, so it just doesn't make sense for the plugin to build the context itself. On top of that plugins can already do this via OnActivity and all the other lifecycle hooks
 
> devs should be interacting with plugins as little as possible, they were designed to be import and forget, they run and do things in the background but they should not be mutated by the dev in 99% of cases 
 
> the Context is the property of the App class, plugins can mutate it in the same way devs can in their handlers, but they do not create the context. If you did want to add some context factory (still strange since it is fully possible today via a route handler like you would with routers like express), it would need to be added to the App class, not a plugin.
 
> tldr, adding this stuff to the plugins would be a breach of separation of concerns and would cross the virtual boundary of responsibilities that the App and Plugins have.
